inst/www/shared/shiny.js

```diff
@@ -1260,106 +1279,801 @@
+            shiftAmount = min - minval;
+          }
+
+          var newvals = [];
+          for (var i=0; i<vals.length; i++) {
+            newvals[i] = vals[i] + shiftAmount;
+          }
+          return newvals;
         }
 
-        if (clickId || hoverId) {
-          $(img).addClass('crosshair');
+        // Set cursor to one of 3 styles. We need to set the cursor on the whole
+        // img instead of the brush dive, because the brush div has
+        // 'pointer-events:none' so that pointer events will pass through to the
+        // img.
+        function setCursorStyle(style) {
+          if (style === 'crosshair') {
+            $el.addClass('crosshair');
+            $el.removeClass('grabbable');
+            $el.removeClass('grabbable');
+
+          } else if (style === 'grabbable') {
+            $el.removeClass('crosshair');
+            $el.addClass('grabbable');
+            $el.removeClass('grabbing');
+
+          } else if (style === 'grabbing') {
+            $el.removeClass('crosshair');
+            $el.removeClass('grabbable');
+            $el.addClass('grabbing');
+          }
         }
+
+        function sendBrushInfo() {
+          var bounds = brush.bounds;
+
+          // We're in a new or reset state
+          if (isNaN(bounds.xmin)) {
+            exports.onInputChange(inputId, null);
+            return;
+          }
+
+          // Transform coordinates of brush to data space
+          var min = offsetToScaledCoords({ x: bounds.xmin, y: bounds.ymin }, opts.brushClip);
+          var max = offsetToScaledCoords({ x: bounds.xmax, y: bounds.ymax }, opts.brushClip);
+
+          // Because the x and y directions of the pixel space may differ from
+          // the x and y directions of the data space, we need to recalculate
+          // the min and max.
+          var coords = {
+            xmin: Math.min(min.x, max.x),
+            xmax: Math.max(min.x, max.x),
+            ymin: Math.min(min.y, max.y),
+            ymax: Math.max(min.y, max.y),
+            '.nonce': Math.random()
+          };
+
+          // Send data to server
+          exports.onInputChange(inputId, coords);
+        }
+
+        var brushInfoSender;
+        if (opts.brushDelayType === 'throttle') {
+          brushInfoSender = new Throttler(null, sendBrushInfo, opts.brushDelay);
+        } else {
+          brushInfoSender = new Debouncer(null, sendBrushInfo, opts.brushDelay);
+        }
+
+        function mousedown(e) {
+          // This can happen when mousedown inside the graphic, then mouseup
+          // outside, then mousedown inside. Just ignore the second
+          // mousedown.
+          if (brush.brushing || brush.dragging) return;
+
+          // Listen for left mouse button only
+          if (e.which !== 1) return;
+
+          var offset = mouseOffset(e);
+
+          // Ignore mousedown events outside of plotting region
+          if (opts.brushClip && !isInPlottingRegion(offset)) return;
```

Comment:
If this is doing what I think it is, I think it would be better if the offset was moved into the plotting region (instead of flat out ignored). So if you start a brushing operation above and to the left of the top-left corner, the brush would start at the top-left corner.